File: examples/showcase/jira_fronting/custom-code-bundles/jira_fronting_typescript/src/runtime/backend-adapter.ts
import type { BackendInvocationPlan, GeneratedCapabilityRuntimeMetadata } from "../generated/runtime-target.js";

export type GeneratedBackendInvocationContext = {
  rootPrincipal?: string;
};

export interface GeneratedBackendAdapter {
  execute(capability: GeneratedCapabilityRuntimeMetadata, plan: BackendInvocationPlan, adapterInput: Record<string, unknown>, context: GeneratedBackendInvocationContext): Promise<Record<string, unknown>>;
}

type JiraConfig = {
  baseUrl: string;
  email: string;
  token: string;
};

function jiraConfig(): JiraConfig | null {
  const baseUrl = (process.env.JIRA_BASE_URL ?? "").replace(/\/+$/, "");
  const email = process.env.JIRA_EMAIL ?? "";
  const token = process.env.JIRA_API_TOKEN ?? "";
  if (!baseUrl || !email || !token) return null;
  return { baseUrl, email, token };
}

function authHeaders(config: JiraConfig): Record<string, string> {
  return {
    Accept: "application/json",
    Authorization: `Basic ${Buffer.from(`${config.email}:${config.token}`).toString("base64")}`,
  };
}

async function jiraJson(config: JiraConfig, method: string, path: string, query?: Record<string, string>, body?: Record<string, unknown>): Promise<Record<string, unknown>> {
  const url = new URL(`${config.baseUrl}${path}`);
  for (const [key, value] of Object.entries(query ?? {})) url.searchParams.set(key, value);
  const response = await fetch(url, {
    method,
    headers: body ? { ...authHeaders(config), "Content-Type": "application/json" } : authHeaders(config),
    body: body ? JSON.stringify(body) : undefined,
  });
  const text = await response.text();
  const payload = text ? JSON.parse(text) as Record<string, unknown> : {};
  if (!response.ok) return { error: "jira_http_error", status: response.status, detail: payload };
  return payload;
}

function boundedLimit(value: unknown, defaultValue = 25, maximum = 50): number {
  const parsed = Number.parseInt(String(value ?? ""), 10);
  const limit = Number.isFinite(parsed) ? parsed : defaultValue;
  return Math.max(1, Math.min(limit, maximum));
}

function safeJqlValue(value: unknown): string {
  return String(value ?? "").trim().replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}

function listValue(value: unknown): string[] {
  const raw = Array.isArray(value) ? value : typeof value === "string" && value.trim() ? value.split(",") : [];
  const result: string[] = [];
  for (const item of raw) {
    const text = String(item).trim();
    if (text && !result.includes(text)) result.push(text);
  }
  return result;
}

function safeLabel(value: unknown): string {
  return String(value ?? "").trim().toLowerCase().replace(/[^a-z0-9_.-]+/g, "-").replace(/^-+|-+$/g, "").slice(0, 255);
}

function labels(value: unknown): string[] {
  return listValue(value).map(safeLabel).filter(Boolean);
}

function adfDoc(text: unknown): Record<string, unknown> {
  return {
    type: "doc",
    version: 1,
    content: [{ type: "paragraph", content: [{ type: "text", text: String(text ?? "").trim() }] }],
  };
}

function plainTextFromAdf(value: unknown): string {
  const parts: string[] = [];
  const walk = (node: unknown): void => {
    if (Array.isArray(node)) {
      for (const child of node) walk(child);
      return;
    }
    if (!node || typeof node !== "object") return;
    const record = node as Record<string, unknown>;
    if (record.type === "text" && typeof record.text === "string") parts.push(record.text);
    walk(record.content);
  };
  walk(value);
  return parts.map((part) => part.trim()).filter(Boolean).join(" ");
}

function issueSummary(issue: Record<string, unknown>): Record<string, unknown> {
  const fields = (issue.fields ?? {}) as Record<string, unknown>;
  return {
    key: issue.key,
    summary: fields.summary,
    status: (fields.status as Record<string, unknown> | undefined)?.name,
    issue_type: (fields.issuetype as Record<string, unknown> | undefined)?.name,
    project_key: (fields.project as Record<string, unknown> | undefined)?.key,
    assignee: (fields.assignee as Record<string, unknown> | undefined)?.displayName,
    priority: (fields.priority as Record<string, unknown> | undefined)?.name,
  };
}

function issueQueryJql(projectKey: string, queryText: string): string {
  let jql = `project = "${safeJqlValue(projectKey)}"`;
  if (queryText) jql += ` AND text ~ "${safeJqlValue(queryText)}"`;
  return `${jql} ORDER BY updated DESC`;
}

async function searchIssues(config: JiraConfig, jql: string, limit: number, fields: string): Promise<Record<string, unknown>> {
  return jiraJson(config, "GET", "/rest/api/3/search/jql", { jql, maxResults: String(limit), fields });
}

function priorityForSeverity(value: unknown): string {
  const severity = String(value ?? "").toLowerCase();
  if (severity === "sev1" || severity === "sev2") return "High";
  if (severity === "sev4") return "Low";
  return "Medium";
}

function previewResult(capability: GeneratedCapabilityRuntimeMetadata, plan: BackendInvocationPlan, action: string, request: Record<string, unknown>, metadata: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    execution_status: "prepared",
    capability_id: capability.capability_id,
    selected_backend: plan.selected_binding,
    semantic_input: plan.semantic_input,
    approval_required: capability.operation_type === "approval_gated" || capability.execution_posture === "prepare_only",
    mutation_performed: false,
    jira_action: action,
    jira_request_preview: request,
    jira_metadata: metadata,
    note: "Prepared a governed Jira request preview. No Jira mutation was performed.",
  };
}

function backendError(capability: GeneratedCapabilityRuntimeMetadata, plan: BackendInvocationPlan, payload: Record<string, unknown>): Record<string, unknown> {
  return { execution_status: "backend_error", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, jira_error: payload };
}

export function createDefaultBackendAdapter(): GeneratedBackendAdapter {
  return {
    async execute(capability, plan, adapterInput, _context) {
      if (plan.unresolved_required_backend_inputs.length > 0) {
        return { execution_status: "backend_input_incomplete", capability_id: capability.capability_id, backend_input_contract: plan.backend_input_contract, unresolved_required_backend_inputs: plan.unresolved_required_backend_inputs };
      }
      const config = jiraConfig();
      const params = adapterInput;
      switch (capability.capability_id) {
        case "jira.backlog.search_context": {
          const projectKey = String(params.project_key ?? "").trim();
          const queryText = String(params.query ?? "").trim();
          let jql = issueQueryJql(projectKey, queryText);
          if (params.issue_type) jql += ` AND issuetype = "${safeJqlValue(params.issue_type)}"`;
          if (params.status) jql += ` AND status = "${safeJqlValue(params.status)}"`;
          if (!config) return { execution_status: "backend_not_configured", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, jql_preview: jql };
          const payload = await searchIssues(config, jql, boundedLimit(params.limit), "summary,status,issuetype,project,assignee,priority");
          if (payload.error) return backendError(capability, plan, payload);
          const issues = ((payload.issues as Record<string, unknown>[] | undefined) ?? []).map(issueSummary);
          return { execution_status: "completed", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, jql, result: { issues, count: issues.length, is_last: payload.isLast } };
        }
        case "jira.issue.get_context": {
          const issueKey = String(params.issue_key ?? "").trim();
          if (!config) return { execution_status: "backend_not_configured", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, path_preview: `/rest/api/3/issue/${issueKey}` };
          const payload = await jiraJson(config, "GET", `/rest/api/3/issue/${encodeURIComponent(issueKey)}`, { fields: "summary,status,issuetype,project,assignee,priority,description" });
          if (payload.error) return backendError(capability, plan, payload);
          const result = issueSummary(payload);
          const description = (payload.fields as Record<string, unknown> | undefined)?.description;
          if (description) result.description_excerpt = plainTextFromAdf(description).slice(0, 500);
          if (String(params.include_comments ?? "").toLowerCase() === "true") {
            const comments = await jiraJson(config, "GET", `/rest/api/3/issue/${encodeURIComponent(issueKey)}/comment`, { maxResults: "5" });
            if (!comments.error) result.comments = ((comments.comments as Record<string, unknown>[] | undefined) ?? []).map((comment) => ({
              author: (comment.author as Record<string, unknown> | undefined)?.displayName,
              body_excerpt: plainTextFromAdf(comment.body).slice(0, 500),
              created: comment.created,
            }));
          }
          return { execution_status: "completed", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, result };
        }
        case "jira.release_notes.prepare": {
          const projectKey = String(params.project_key ?? "").trim();
          const releaseRef = String(params.release_ref ?? "").trim();
          const audience = String(params.audience ?? "internal").trim();
          let jql = releaseRef.toLowerCase() === "unversioned" ? `project = "${safeJqlValue(projectKey)}" AND fixVersion is EMPTY` : `project = "${safeJqlValue(projectKey)}" AND fixVersion = "${safeJqlValue(releaseRef)}"`;
          if (params.issue_query) jql += ` AND text ~ "${safeJqlValue(params.issue_query)}"`;
          jql += " ORDER BY priority DESC, updated DESC";
          let issues: Record<string, unknown>[] = [];
          if (config) {
            const payload = await searchIssues(config, jql, boundedLimit(params.limit, 20, 50), "summary,status,issuetype,project");
            if (payload.error) return backendError(capability, plan, payload);
            issues = ((payload.issues as Record<string, unknown>[] | undefined) ?? []).map(issueSummary);
          }
          const draft = issues.length ? [`Release ${releaseRef} notes for ${audience}`, "", ...issues.map((issue) => `- ${issue.key}: ${issue.summary} (${issue.status})`)].join("\n") : `Release ${releaseRef} notes for ${audience}\n\nNo matching Jira issues were returned for the bounded query.`;
          return { execution_status: "prepared", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, jql, result: { audience, issue_count: issues.length, issues, draft }, note: "Prepared release notes only. No Jira mutation or publication was performed." };
        }
        case "jira.incident_bug.prepare":
        case "jira.story.prepare": {
          const issueTypeName = capability.capability_id === "jira.incident_bug.prepare" ? "Bug" : "Story";
          const fields: Record<string, unknown> = { project: { key: params.project_key }, issuetype: { name: issueTypeName }, summary: String(params.summary ?? "").trim() };
          if (capability.capability_id === "jira.incident_bug.prepare") {
            fields.description = adfDoc(params.description);
            fields.priority = { name: priorityForSeverity(params.severity) };
          } else {
            fields.description = adfDoc(`Acceptance criteria:\n${params.acceptance_criteria ?? ""}`);
            if (params.priority) fields.priority = { name: String(params.priority).replace(/^./, (ch) => ch.toUpperCase()) };
          }
          const labelValues = labels(params.labels);
          if (labelValues.length) fields.labels = labelValues;
          return previewResult(capability, plan, "create_issue", { method: "POST", path: "/rest/api/3/issue", body: { fields } }, { project_key: params.project_key, requested_issue_type: issueTypeName });
        }
        case "jira.subtask.prepare": {
          const parentIssueKey = String(params.parent_issue_key ?? "").trim();
          const fields: Record<string, unknown> = { parent: { key: parentIssueKey }, issuetype: { name: "Sub-task" }, summary: String(params.summary ?? "").trim(), description: adfDoc(params.description) };
          if (config) {
            const parent = await jiraJson(config, "GET", `/rest/api/3/issue/${encodeURIComponent(parentIssueKey)}`, { fields: "project" });
            const projectKey = ((parent.fields as Record<string, unknown> | undefined)?.project as Record<string, unknown> | undefined)?.key;
            if (projectKey) fields.project = { key: projectKey };
            if (parent.id) fields.parent = { id: parent.id };
          }
          return previewResult(capability, plan, "create_subtask", { method: "POST", path: "/rest/api/3/issue", body: { fields } }, { parent_issue_key: parentIssueKey });
        }
        case "jira.customer_escalation.comment.prepare":
          return previewResult(capability, plan, "add_comment", { method: "POST", path: `/rest/api/3/issue/${params.issue_key}/comment`, body: { body: adfDoc(`[${params.comment_purpose}] ${params.context ?? ""}`), visibility: params.visibility ?? "internal" } }, { issue_key: params.issue_key, visibility: params.visibility ?? "internal", comment_purpose: params.comment_purpose });
        case "jira.workflow_transition.request":
          return previewResult(capability, plan, "transition_issue", { method: "POST", path: `/rest/api/3/issue/${params.issue_key}/transitions`, body: { transition: { id: params.target_status }, ...(params.comment ? { update: { comment: [{ add: { body: adfDoc(params.comment) } }] } } : {}) } }, { issue_key: params.issue_key, target_status: params.target_status });
        case "jira.sprint_move.request":
          return previewResult(capability, plan, "move_issues_to_sprint", { method: "POST", path: `/rest/agile/1.0/sprint/${params.target_sprint}/issue`, body: { issues: listValue(params.issue_keys) } }, { issue_keys: listValue(params.issue_keys), target_sprint: params.target_sprint });
        case "jira.assignee_change.request":
          return previewResult(capability, plan, "assign_issue", { method: "PUT", path: `/rest/api/3/issue/${params.issue_key}/assignee`, body: { accountId: params.assignee_ref } }, { issue_key: params.issue_key, assignee_ref: params.assignee_ref });
        case "jira.issue_link.request":
          return previewResult(capability, plan, "link_issues", { method: "POST", path: "/rest/api/3/issueLink", body: { type: { name: params.link_type }, inwardIssue: { key: params.source_issue_key }, outwardIssue: { key: params.target_issue_key }, comment: { body: adfDoc(params.reason) } } }, { requested_link_type: params.link_type });
        default:
          return { execution_status: "backend_execution_stub", capability_id: capability.capability_id, selected_backend: plan.selected_binding, semantic_input: plan.semantic_input, backend_input_contract: plan.backend_input_contract, note: "No Jira custom handler is registered for this capability." };
      }
    },
  };
}

export const backendAdapter = createDefaultBackendAdapter();

File: examples/showcase/jira_fronting/custom-code-bundles/jira_fronting_go/extensions/backend_adapter.go
package extensions

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strconv"
	"strings"

	"{{ANIP_GO_MODULE_PATH}}/generated"
)

type BackendInvocationContext struct {
	RootPrincipal string
}

type BackendAdapter interface {
	Execute(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, adapterInput map[string]any, context BackendInvocationContext) (map[string]any, error)
}

type jiraConfig struct {
	BaseURL string
	Email   string
	Token   string
}

type jiraBackendAdapter struct{}

func CreateDefaultBackendAdapter() BackendAdapter {
	return jiraBackendAdapter{}
}

func (jiraBackendAdapter) Execute(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, _ BackendInvocationContext) (map[string]any, error) {
	if len(plan.UnresolvedRequiredBackendInputs) > 0 {
		return map[string]any{
			"execution_status": "backend_input_incomplete",
			"capability_id": capability.CapabilityID,
			"backend_input_contract": plan.BackendInputContract,
			"unresolved_required_backend_inputs": plan.UnresolvedRequiredBackendInputs,
		}, nil
	}
	config := readJiraConfig()
	switch capability.CapabilityID {
	case "jira.backlog.search_context":
		return searchBacklog(capability, plan, params, config), nil
	case "jira.issue.get_context":
		return getIssueContext(capability, plan, params, config), nil
	case "jira.release_notes.prepare":
		return prepareReleaseNotes(capability, plan, params, config), nil
	case "jira.incident_bug.prepare":
		return prepareIssueCreate(capability, plan, params, "Bug"), nil
	case "jira.story.prepare":
		return prepareIssueCreate(capability, plan, params, "Story"), nil
	case "jira.subtask.prepare":
		return prepareSubtask(capability, plan, params, config), nil
	case "jira.customer_escalation.comment.prepare":
		return previewResult(capability, plan, "add_comment", map[string]any{"method": "POST", "path": "/rest/api/3/issue/" + text(params["issue_key"]) + "/comment", "body": map[string]any{"body": adfDoc("["+text(params["comment_purpose"])+"] "+text(params["context"])), "visibility": adapterFirstNonEmpty(text(params["visibility"]), "internal")}}, map[string]any{"issue_key": params["issue_key"], "visibility": adapterFirstNonEmpty(text(params["visibility"]), "internal"), "comment_purpose": params["comment_purpose"]}), nil
	case "jira.workflow_transition.request":
		body := map[string]any{"transition": map[string]any{"id": params["target_status"]}}
		if comment := text(params["comment"]); comment != "" {
			body["update"] = map[string]any{"comment": []any{map[string]any{"add": map[string]any{"body": adfDoc(comment)}}}}
		}
		return previewResult(capability, plan, "transition_issue", map[string]any{"method": "POST", "path": "/rest/api/3/issue/" + text(params["issue_key"]) + "/transitions", "body": body}, map[string]any{"issue_key": params["issue_key"], "target_status": params["target_status"]}), nil
	case "jira.sprint_move.request":
		issues := listValue(params["issue_keys"])
		return previewResult(capability, plan, "move_issues_to_sprint", map[string]any{"method": "POST", "path": "/rest/agile/1.0/sprint/" + text(params["target_sprint"]) + "/issue", "body": map[string]any{"issues": issues}}, map[string]any{"issue_keys": issues, "target_sprint": params["target_sprint"]}), nil
	case "jira.assignee_change.request":
		return previewResult(capability, plan, "assign_issue", map[string]any{"method": "PUT", "path": "/rest/api/3/issue/" + text(params["issue_key"]) + "/assignee", "body": map[string]any{"accountId": params["assignee_ref"]}}, map[string]any{"issue_key": params["issue_key"], "assignee_ref": params["assignee_ref"]}), nil
	case "jira.issue_link.request":
		return previewResult(capability, plan, "link_issues", map[string]any{"method": "POST", "path": "/rest/api/3/issueLink", "body": map[string]any{"type": map[string]any{"name": params["link_type"]}, "inwardIssue": map[string]any{"key": params["source_issue_key"]}, "outwardIssue": map[string]any{"key": params["target_issue_key"]}, "comment": map[string]any{"body": adfDoc(params["reason"])}}}, map[string]any{"requested_link_type": params["link_type"]}), nil
	default:
		return map[string]any{"execution_status": "backend_execution_stub", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "backend_input_contract": plan.BackendInputContract, "note": "No Jira custom handler is registered for this capability."}, nil
	}
}

func readJiraConfig() *jiraConfig {
	baseURL := strings.TrimRight(os.Getenv("JIRA_BASE_URL"), "/")
	email := os.Getenv("JIRA_EMAIL")
	token := os.Getenv("JIRA_API_TOKEN")
	if baseURL == "" || email == "" || token == "" {
		return nil
	}
	return &jiraConfig{BaseURL: baseURL, Email: email, Token: token}
}

func jiraJSON(config *jiraConfig, method, path string, query map[string]string, body map[string]any) map[string]any {
	if config == nil {
		return map[string]any{"error": "jira_not_configured"}
	}
	target, _ := url.Parse(config.BaseURL + path)
	values := target.Query()
	for key, value := range query {
		values.Set(key, value)
	}
	target.RawQuery = values.Encode()
	var requestBody *strings.Reader
	if body != nil {
		encoded, _ := json.Marshal(body)
		requestBody = strings.NewReader(string(encoded))
	} else {
		requestBody = strings.NewReader("")
	}
	request, _ := http.NewRequest(method, target.String(), requestBody)
	request.Header.Set("Accept", "application/json")
	request.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(config.Email+":"+config.Token)))
	if body != nil {
		request.Header.Set("Content-Type", "application/json")
	}
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		return map[string]any{"error": "jira_http_error", "detail": err.Error()}
	}
	defer response.Body.Close()
	var payload map[string]any
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		payload = map[string]any{}
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return map[string]any{"error": "jira_http_error", "status": response.StatusCode, "detail": payload}
	}
	return payload
}

func searchBacklog(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, config *jiraConfig) map[string]any {
	projectKey := text(params["project_key"])
	queryText := text(params["query"])
	jql := issueQueryJQL(projectKey, queryText)
	if issueType := text(params["issue_type"]); issueType != "" {
		jql += ` AND issuetype = "` + safeJQLValue(issueType) + `"`
	}
	if status := text(params["status"]); status != "" {
		jql += ` AND status = "` + safeJQLValue(status) + `"`
	}
	if config == nil {
		return map[string]any{"execution_status": "backend_not_configured", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "jql_preview": jql}
	}
	payload := searchIssues(config, jql, boundedLimit(params["limit"], 25, 50), "summary,status,issuetype,project,assignee,priority")
	if payload["error"] != nil {
		return backendError(capability, plan, payload)
	}
	issues := summarizeIssues(payload["issues"])
	return map[string]any{"execution_status": "completed", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "jql": jql, "result": map[string]any{"issues": issues, "count": len(issues), "is_last": payload["isLast"]}}
}

func getIssueContext(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, config *jiraConfig) map[string]any {
	issueKey := text(params["issue_key"])
	if config == nil {
		return map[string]any{"execution_status": "backend_not_configured", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "path_preview": "/rest/api/3/issue/" + issueKey}
	}
	payload := jiraJSON(config, "GET", "/rest/api/3/issue/"+url.PathEscape(issueKey), map[string]string{"fields": "summary,status,issuetype,project,assignee,priority,description"}, nil)
	if payload["error"] != nil {
		return backendError(capability, plan, payload)
	}
	result := issueSummary(payload)
	return map[string]any{"execution_status": "completed", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "result": result}
}

func prepareReleaseNotes(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, config *jiraConfig) map[string]any {
	projectKey := text(params["project_key"])
	releaseRef := text(params["release_ref"])
	audience := adapterFirstNonEmpty(text(params["audience"]), "internal")
	var jql string
	if strings.EqualFold(releaseRef, "unversioned") {
		jql = `project = "` + safeJQLValue(projectKey) + `" AND fixVersion is EMPTY`
	} else {
		jql = `project = "` + safeJQLValue(projectKey) + `" AND fixVersion = "` + safeJQLValue(releaseRef) + `"`
	}
	if issueQuery := text(params["issue_query"]); issueQuery != "" {
		jql += ` AND text ~ "` + safeJQLValue(issueQuery) + `"`
	}
	jql += " ORDER BY priority DESC, updated DESC"
	issues := []map[string]any{}
	if config != nil {
		payload := searchIssues(config, jql, boundedLimit(params["limit"], 20, 50), "summary,status,issuetype,project")
		if payload["error"] != nil {
			return backendError(capability, plan, payload)
		}
		issues = summarizeIssues(payload["issues"])
	}
	return map[string]any{"execution_status": "prepared", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "jql": jql, "result": map[string]any{"audience": audience, "issue_count": len(issues), "issues": issues, "draft": releaseNoteDraft(audience, releaseRef, issues)}, "note": "Prepared release notes only. No Jira mutation or publication was performed."}
}

func prepareIssueCreate(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, issueTypeName string) map[string]any {
	fields := map[string]any{"project": map[string]any{"key": params["project_key"]}, "issuetype": map[string]any{"name": issueTypeName}, "summary": text(params["summary"])}
	if capability.CapabilityID == "jira.incident_bug.prepare" {
		fields["description"] = adfDoc(params["description"])
		fields["priority"] = map[string]any{"name": priorityForSeverity(params["severity"])}
	} else {
		fields["description"] = adfDoc("Acceptance criteria:\n" + text(params["acceptance_criteria"]))
		if priority := text(params["priority"]); priority != "" {
			fields["priority"] = map[string]any{"name": strings.Title(priority)}
		}
	}
	if labels := labels(params["labels"]); len(labels) > 0 {
		fields["labels"] = labels
	}
	return previewResult(capability, plan, "create_issue", map[string]any{"method": "POST", "path": "/rest/api/3/issue", "body": map[string]any{"fields": fields}}, map[string]any{"project_key": params["project_key"], "requested_issue_type": issueTypeName})
}

func prepareSubtask(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, params map[string]any, config *jiraConfig) map[string]any {
	parentIssueKey := text(params["parent_issue_key"])
	fields := map[string]any{"parent": map[string]any{"key": parentIssueKey}, "issuetype": map[string]any{"name": "Sub-task"}, "summary": text(params["summary"]), "description": adfDoc(params["description"])}
	if config != nil {
		parent := jiraJSON(config, "GET", "/rest/api/3/issue/"+url.PathEscape(parentIssueKey), map[string]string{"fields": "project"}, nil)
		if parent["error"] == nil {
			if projectKey := nestedString(parent, "fields", "project", "key"); projectKey != "" {
				fields["project"] = map[string]any{"key": projectKey}
			}
			if id := text(parent["id"]); id != "" {
				fields["parent"] = map[string]any{"id": id}
			}
		}
	}
	return previewResult(capability, plan, "create_subtask", map[string]any{"method": "POST", "path": "/rest/api/3/issue", "body": map[string]any{"fields": fields}}, map[string]any{"parent_issue_key": parentIssueKey})
}

func previewResult(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, action string, request map[string]any, metadata map[string]any) map[string]any {
	return map[string]any{"execution_status": "prepared", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "approval_required": capability.OperationType == "approval_gated" || capability.ExecutionPosture == "prepare_only", "mutation_performed": false, "jira_action": action, "jira_request_preview": request, "jira_metadata": metadata, "note": "Prepared a governed Jira request preview. No Jira mutation was performed."}
}

func backendError(capability generated.GeneratedCapabilityRuntimeMetadata, plan generated.BackendInvocationPlan, payload map[string]any) map[string]any {
	return map[string]any{"execution_status": "backend_error", "capability_id": capability.CapabilityID, "selected_backend": plan.SelectedBinding, "semantic_input": plan.SemanticInput, "jira_error": payload}
}

func searchIssues(config *jiraConfig, jql string, limit int, fields string) map[string]any {
	return jiraJSON(config, "GET", "/rest/api/3/search/jql", map[string]string{"jql": jql, "maxResults": strconv.Itoa(limit), "fields": fields}, nil)
}

func issueQueryJQL(projectKey, queryText string) string {
	jql := `project = "` + safeJQLValue(projectKey) + `"`
	if queryText != "" {
		jql += ` AND text ~ "` + safeJQLValue(queryText) + `"`
	}
	return jql + " ORDER BY updated DESC"
}

func issueSummary(issue map[string]any) map[string]any {
	return map[string]any{"key": issue["key"], "summary": nestedString(issue, "fields", "summary"), "status": nestedString(issue, "fields", "status", "name"), "issue_type": nestedString(issue, "fields", "issuetype", "name"), "project_key": nestedString(issue, "fields", "project", "key"), "assignee": nestedString(issue, "fields", "assignee", "displayName"), "priority": nestedString(issue, "fields", "priority", "name")}
}

func summarizeIssues(value any) []map[string]any {
	items, _ := value.([]any)
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		if issue, ok := item.(map[string]any); ok {
			result = append(result, issueSummary(issue))
		}
	}
	return result
}

func adfDoc(value any) map[string]any {
	return map[string]any{"type": "doc", "version": 1, "content": []any{map[string]any{"type": "paragraph", "content": []any{map[string]any{"type": "text", "text": text(value)}}}}}
}

func releaseNoteDraft(audience, releaseRef string, issues []map[string]any) string {
	heading := fmt.Sprintf("Release %s notes for %s", releaseRef, audience)
	if len(issues) == 0 {
		return heading + "\n\nNo matching Jira issues were returned for the bounded query."
	}
	lines := []string{heading, ""}
	for _, issue := range issues {
		lines = append(lines, fmt.Sprintf("- %s: %s (%s)", issue["key"], issue["summary"], issue["status"]))
	}
	return strings.Join(lines, "\n")
}

func boundedLimit(value any, defaultValue, maximum int) int {
	limit, err := strconv.Atoi(text(value))
	if err != nil {
		limit = defaultValue
	}
	if limit < 1 {
		return 1
	}
	if limit > maximum {
		return maximum
	}
	return limit
}

func safeJQLValue(value any) string {
	return strings.ReplaceAll(strings.ReplaceAll(text(value), `\`, `\\`), `"`, `\"`)
}

func text(value any) string {
	if value == nil {
		return ""
	}
	return strings.TrimSpace(fmt.Sprint(value))
}

func listValue(value any) []string {
	var raw []any
	switch typed := value.(type) {
	case []any:
		raw = typed
	case []string:
		for _, item := range typed {
			raw = append(raw, item)
		}
	case string:
		for _, item := range strings.Split(typed, ",") {
			raw = append(raw, item)
		}
	}
	result := []string{}
	for _, item := range raw {
		if candidate := text(item); candidate != "" && !contains(result, candidate) {
			result = append(result, candidate)
		}
	}
	return result
}

func labels(value any) []string {
	allowed := regexp.MustCompile(`[^a-z0-9_.-]+`)
	result := []string{}
	for _, item := range listValue(value) {
		label := strings.Trim(allowed.ReplaceAllString(strings.ToLower(item), "-"), "-")
		if label != "" {
			result = append(result, label)
		}
	}
	return result
}

func priorityForSeverity(value any) string {
	switch strings.ToLower(text(value)) {
	case "sev1", "sev2":
		return "High"
	case "sev4":
		return "Low"
	default:
		return "Medium"
	}
}

func nestedString(value map[string]any, path ...string) string {
	var current any = value
	for _, key := range path {
		record, ok := current.(map[string]any)
		if !ok {
			return ""
		}
		current = record[key]
	}
	return text(current)
}

func adapterFirstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func contains(values []string, candidate string) bool {
	for _, value := range values {
		if value == candidate {
			return true
		}
	}
	return false
}

var BackendAdapterInstance = CreateDefaultBackendAdapter()
